TestModels/LocalService/test/SimpleLocalServiceTest.dfy

```diff
@@ -8,14 +8,35 @@ module SimpleLocalServiceTest {
   import Types = SimpleLocalServiceTypes
   import opened Wrappers
 
-  method TestClient(config: Types.SimpleLocalServiceConfig)
+  method TestClientDirect(client: Types.ISimpleLocalServiceClient)
+    requires client.ValidState()
+    modifies client.Modifies
+    ensures client.ValidState()
   {
-    var client :- expect SimpleLocalService.SimpleLocalService(config);
+    var ret :- expect client.HelloWorld(Types.HelloWorldInput());
+    expect ret.greeting == "Hi there!";
+  }
+
+  method TestClientSelfReference(client: Types.ISimpleLocalServiceClient)
+    requires client.ValidState()
+    modifies client.Modifies
+    ensures client.ValidState()
+  {
+    // This is a flat out lie and directly contradicts client.ValidState().
+    // But otherwise it's not possible to call other client methods like this.
+    // Since this test model is only using self-reference in order to
+    // avoid extra dependencies, and self-reference isn't that useful a use case in reality,
+    // lying to get around the specification restrictions in a test is not terribly risky.
+    assume {:axiom} client.Modifies !! {client.History};
```

Comment:
My position on tests in Dafny is that sometimes it is better to outright lie to verify the behavior you want to test.

This is because Dafny verification is a kind of test, and so we write the specifications so we don't need to test. But if we want to confirm the spec, we need some lies.

Reply:
I'd much rather not lie since you can't actually reliably bound the lie in general - I'm only comfortable here because it's only test code etc.

To be clearer the actual problem is that the generated spec for `SelfReflection` includes `requires input.self.Modifies !! {History}`, which is impossible when `input.self` is the client itself.

The good news is I realized I could avoid the issue by using a second client - see latest commit.